### DISCUSSION TITLE
Defaults: `memory_limit` vs `upload_max_filesize`

### DISCUSSION THREAD

**Discussion by Leksat** [uselagoon/lagoon#3485] [General]
From https://docs.lagoon.sh/docker-images/php-fpm/#included-php-config
> | Value | Details |
> | :--- | :--- |
> | `memory_limit = 400M` | For big PHP projects \(changeable via `PHP_MEMORY_LIMIT`\). |
> | `upload_max_filesize = 2048M` | For big file uploads. |

But according to PHP docs
- `memory_limit` should be larger than `post_max_size` ([link](https://www.php.net/manual/en/ini.core.php#ini.post-max-size))
- `post_max_size` must be larger than `upload_max_filesize` ([link](https://www.php.net/manual/en/ini.core.php#ini.upload-max-filesize)).

So `memory_limit > post_max_size > upload_max_filesize`. Or at least they should be equal. But [current defaults](https://github.com/uselagoon/lagoon-images/blob/b7495451d8e62ac61b7c107cf047059604b274aa/images/php-fpm/00-lagoon-php.ini.tpl#L10) are different. So files more that `400M` can't be uploaded.

**Comment by seanhamlin** [comment]
See https://stackoverflow.com/a/3651195/169882, the TLDR is that `memory_limit` **does not** need to be > `post_max_size`. 

Checking a brand new site install on `23.06.0` images:

```
cli-drupal:/app$ php -i | grep -E 'upload_max_filesize|post_max_size'
post_max_size => 2048M => 2048M
upload_max_filesize => 2048M => 2048M
```

So `post_max_size` == `upload_max_filesize` @ 2 GB.

Pretty sure customers have uploaded large files through the web, and this is why the 2GB is in place by default.

**Reply by Leksat** [reply]
Hmm... Our experience is different. PHP died with out of memory on file upload until we increased `memory_limit`.

**Reply by Leksat** [reply]
Maybe it has something to do with the Drupal setup 🤔

**Reply by seanhamlin** [reply]
I just tried on a vanilla Drupal 10.1 on Lagoon, with a 560MB file, and indeed it does die

```
NOTICE: PHP message: PHP Fatal error:  Allowed memory size of 419430400 bytes exhausted (tried to allocate 536870944 bytes) in /app/web/core/lib/Drupal/Core/File/FileSystem.php on line 76
```

Perhaps Drupal needs PHP's memory to be inline with the max file to be uploaded? I wonder if this is due to the fancy AJAX that is part of the file upload screen (aka, this is not just a boring multi-part form POST)?

<img width="1018" alt="image" src="https://github.com/uselagoon/lagoon/assets/10605846/cc5293c9-e02a-464e-b00b-a72d0b2f25d3">

After bumping the PHP max memory to 800MB

<img width="566" alt="image" src="https://github.com/uselagoon/lagoon/assets/10605846/55924d79-75ef-42f1-9d6c-50fe7d21f47f">

It works.

So I think this is potentially a quirk of Drupal's AJAX handling that requires in memory processing of uploads. I wonder what a vanilla form with a PHP handler would do here?

Do you think Lagoon would be better to have `post_max_size` == `upload_max_filesize` == `400MB`?

Potentially this discussion should be an issue I think?

**Reply by Leksat** [reply]
> Do you think Lagoon would be better to have `post_max_size` == `upload_max_filesize` == `400MB`?

At least there will be no confusion.

BTW, you could try disabling JS to see if it's caused by AJAX. Upload form should still work without JS.

**Reply by seanhamlin** [reply]
Looping in @tobybellwood

**Reply by steveworley** [reply]
Technically the memory limit doesn't impact the max upload; it's only because Drupal is trying to load the file into memory during the upload. PHP typically streams the file so it's not loaded fully into memory and would work this way. I suspect that your assertion of Drupal AJAX being the problem.

I think ideally they would be the same values to reduce ambiguity.

**Reply by Leksat** [reply]
I've found a Drupal core issue: https://www.drupal.org/project/drupal/issues/3366403
